dashboard/src/components/DataRecordingContainer.js

```diff
@@ -0,0 +1,147 @@
+const CHART_PLOT_DURATION = 30 * 1000;
+const MINIMUM_DATA_AGE = 500;
+const MAXIMUM_DATA_AGE = CHART_PLOT_DURATION + (2 * MINIMUM_DATA_AGE);
+const MINIMUM_DATA_COUNT = 1;
+const MAXIMUM_DATA_COUNT = 5000;
+
+let timestampOffset = 0;
+
+class DataRecordingContainer {
+
+  constructor() {
+    this.dataRecordings = {};
+  }
+
+  get dataRecordings() {
+    return this._dataRecordings;
+  }
+
+  set dataRecordings(dataRecordings) {
+    this._dataRecordings = dataRecordings;
+  }
+
+  getIds() {
+    return Object.keys(this.dataRecordings);
+  }
+
+  getData(id) {
+    if (!(id in this.dataRecordings)) {
+      this.dataRecordings[id] = [];
+    }
+    return this.dataRecordings[id];
+  }
+
+  setData(id, data) {
+    this.dataRecordings[id] = data;
+  }
+
+  addData(id, data) {
+    this.setData(id, this.getData(id).concat(data));
+  }
+
+  addDataRecording(dataRecording) {
+    this.addData(dataRecording.dataId, dataRecording.dataList);
+  }
+
+  getDataValuesInDimension(id, dimension) {
+    // var values = [];
+    let maximumAggregationTimestamp = Date.now() - MINIMUM_DATA_AGE - timestampOffset;
+    /* this.getData(id)
+      .filter(data => data.aggregationTimestamp < maximumAggregationTimestamp)
+      .forEach(data => {
+        var value;
+ 
+        if (data.hasOwnProperty('value')) {
+          value = data.value;
+        } else {
+          var firstValue = data.values[0];
+          if (firstValue instanceof Array) {
+            value = firstValue[dimension];
+          } else {
+            value = data.values[dimension];
+          }
+        }
+ 
+        values.push(value);
+      }); */
+    return this.getData(id)
+      .filter(data => data.aggregationTimestamp < maximumAggregationTimestamp)
+      .map(data => {
+        let value;
+        if (data.hasOwnProperty('value')) {
+          value = data.value;
+        } else {
+          const firstValue = data.values[0];
+          if (firstValue instanceof Array) {
+            value = firstValue[dimension];
+          } else {
+            value = data.values[dimension];
+          }
+        }
+        return value;
+      });
+    // return values;
+  }
+
+  getDataTimestamps(id) {
+    let timestamps = [];
+    let maximumAggregationTimestamp = Date.now() - MINIMUM_DATA_AGE - timestampOffset;
+    this.getData(id)
+      .filter(data => data.aggregationTimestamp < maximumAggregationTimestamp)
+      .forEach(data => timestamps.push(data.aggregationTimestamp));
+    return timestamps;
+  }
+
+  getDimensions(id) {
+    let data = this.getData(id);
+    if (data.length === 0) {
+      return 0;
+    }
+    let firstData = data[0];
+    if (firstData.hasOwnProperty('value')) {
+      return 1;
+    } else {
```

Comment:
if you return in the if-block it is good practice to not wrap the rest in an else-block. (guard clause)

Reply:
`    if (firstData.hasOwnProperty('value')) {
      return 1;
    } else {
      const firstValue = firstData.values[0];
      if (firstValue instanceof Array) {
        return firstData.values.length * firstValue.length;
      } else {
        return firstData.values.length;
      }
    }`

is this what you mean?:

`    if (firstData.hasOwnProperty('value')) {
      return 1;
    } if (!firstData.hasOwnProperty('value')) {
      const firstValue = firstData.values[0];
      if (firstValue instanceof Array) {
        return firstData.values.length * firstValue.length;
      } else {
        return firstData.values.length;
      }
    }`

Reply:
i mean instead of:
```
if (firstData.hasOwnProperty('value')) {
  return 1;
} else {
  let firstValue = firstData.values[0];
  if (firstValue instanceof Array) {
    return firstData.values.length * firstValue.length;
  } else {
    return firstData.values.length;
  }
}
```
do this:
```
if (firstData.hasOwnProperty('value')) {
  return 1;
}
let firstValue = firstData.values[0];
if (firstValue instanceof Array) {
  return firstData.values.length * firstValue.length;
}
return firstData.values.length;
```